Fixed cmake macro for xacro in groovy

yamokosk opened the pull request:
- Added error handling to execute_process
- Fixed execute_process call to actually convert xacros to urdf

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/ros/xacro/12)

<!-- Reviewable:end -->

Commit:
- Added error handling to execute_process
- Fixed execute_process call to actually convert xacros to urdf

wjwwood commented on cmake/xacro.cmake:
```diff
@@ -3,15 +3,20 @@ set(_xacro_py ${xacro_PACKAGE_PATH}/xacro.py)
-  execute_process(COMMAND ${_xacro_py} --deps ${input}
-    ERROR_VARIABLE _xacro_err_ignore
-    OUTPUT_VARIABLE _xacro_deps_result
+  execute_process(COMMAND ${_xacro_py} ${input}
```
Can you explain why you changed the command here?

wjwwood commented on cmake/xacro.cmake:
```diff
@@ -3,15 +3,20 @@ set(_xacro_py ${xacro_PACKAGE_PATH}/xacro.py)
+    message(SEND_ERROR "Error occured while trying to execute xacro: ${_xacro_err}")
+  else (_xacro_err)
+    #message(STATUS " - Xacro processing results: ${_xacro_deps_result}")
+    separate_arguments(_xacro_deps_result)
```
I think the way you changed the `execute_process` macro call on line 6 above means this variable will not be set...

wjwwood commented:
Can you explain a little bit more about what these changes are supposed to do, especially what was not working before and how these changes address that problem?

yamokosk commented:
Its been a while since I worked on this, but fundamentally I think the use of CMake's execute_process command was incorrect. Very well could be user error on my part, but as far as I can tell, the current xacro cmake macros produce no output - at least the way I was using them. My testcase:

I have very simple but broken xacro:

```
$ more sample.xacro 
<xacro:macro name="generate_tag" params="msg">
    <message> ${msg} </message>
</xacro:macro>

<xacro:generate_tag msg="Hello World!" />
<xacro:generate_tag msg="Hello Universe!" />
```

If I run xacro on the command line, I correctly get:

```
$ rosrun xacro xacro.py sample.xacro 
Expat parsing error.  Check that:
 - Your XML is correctly formed
 - You have the xacro xmlns declaration: xmlns:xacro="http://www.ros.org/wiki/xacro"

Traceback (most recent call last):
  File "/opt/ros/groovy/stacks/xacro/xacro.py", line 35, in <module>
    xacro.main()
  File "/opt/ros/groovy/stacks/xacro/src/xacro.py", line 542, in main
    doc = parse(f)
  File "/usr/lib/python2.7/xml/dom/minidom.py", line 1920, in parse
    return expatbuilder.parse(file)
  File "/usr/lib/python2.7/xml/dom/expatbuilder.py", line 928, in parse
    result = builder.parseFile(file)
  File "/usr/lib/python2.7/xml/dom/expatbuilder.py", line 207, in parseFile
    parser.Parse(buffer, 0)
xml.parsers.expat.ExpatError: unbound prefix: line 1, column 0
```

And a CMakeLists.txt to convert that xacro into a urdf:

```
$ more CMakeLists.txt 
cmake_minimum_required(VERSION 2.4.6)
include($ENV{ROS_ROOT}/core/rosbuild/rosbuild.cmake)

rosbuild_init()

#set the default path for built executables to the "bin" directory
set(EXECUTABLE_OUTPUT_PATH ${PROJECT_SOURCE_DIR}/bin)
#set the default path for built libraries to the "lib" directory
set(LIBRARY_OUTPUT_PATH ${PROJECT_SOURCE_DIR}/lib)

# Try to include xacro macros
#find_package(xacro REQUIRED) # FAILS
# Workaround
rosbuild_find_ros_package(xacro)
include(${xacro_PACKAGE_PATH}/cmake/xacro.cmake)

xacro_add_xacro_file(${PROJECT_SOURCE_DIR}/sample.xacro ${PROJECT_SOURCE_DIR}/sample.urdf)
```

(Note that the example on the wiki uses "find_package(xacro)". For whatever reason, that does not work for me. So I had to go more manual with the "rosbuild_find_ros_package/include" calls. Maybe another problem?). Run rosmake, 

```
$ rosmake
...
[rosmake-0] Starting >>> xacro [ make ]                                                                                                                                                        
[rosmake-0] Finished <<< xacro ROS_NOBUILD in package xacro                                                                                                                                    
[rosmake-0] Starting >>> xacro_cmake_bug [ make ]                                                                                                                                              
[rosmake-0] Finished <<< xacro_cmake_bug [PASS] [ 5.47 seconds ]  
...
```

Everything seems to run fine, however no sample.urdf is created (obviously) but even worse, it just quietly fails.

If I fix the xacro file,

```
<robot xmlns:xacro="http://ros.org/wiki/xacro" name="example">
    <xacro:macro name="generate_tag" params="msg">
        <message> ${msg} </message>
    </xacro:macro>

    <xacro:generate_tag msg="Hello World!" />
    <xacro:generate_tag msg="Hello Universe!" />
</robot>
```

Re-run the command line, it converts/parses fine,

```
$ rosrun xacro xacro.py sample.xacro 
'<?xml version="1.0" ?>
<!-- =================================================================================== -->
<!-- |    This document was autogenerated by xacro from sample.xacro                   | -->
<!-- |    EDITING THIS FILE BY HAND IS NOT RECOMMENDED                                 | -->
<!-- =================================================================================== -->
<robot name="example" xmlns:xacro="http://ros.org/wiki/xacro">
  <message> Hello World! </message>
  <message> Hello Universe! </message>
</robot>
```

But then if I re-run rosmake, still no sample.urdf gets created. Now if you do all that above with my changes, everything works. First, the bad xacro file:

```
$ rosmake
...
[rosmake-1] Starting >>> xacro [ make ]                                                                                                                                                        
[rosmake-1] Finished <<< xacro [PASS] [ 5.16 seconds ]      
[rosmake-2] Starting >>> xacro_cmake_bug [ make ]                                                                                                                                              
[ rosmake ] Last 40 linescro_cmake_bug: 5.4 sec ]                                                                                                                   [ 1 Active 30/31 Complete ]
{-------------------------------------------------------------------------------
  -- Using CATKIN_TEST_RESULTS_DIR: /home/yamokosk/groovy_workspace_old/xacro_cmake_bug/build/test_results
  -- Found gtest: gtests will be built
  -- catkin 0.5.77
  [rosbuild] Including /opt/ros/groovy/share/roslisp/rosbuild/roslisp.cmake
  [rosbuild] Including /opt/ros/groovy/share/roscpp/rosbuild/roscpp.cmake
  [rosbuild] Including /opt/ros/groovy/share/rospy/rosbuild/rospy.cmake

  [rosbuild] WARNING: the file /home/yamokosk/groovy_workspace_old/xacro/cmake/xacro.cmake is being included automatically.  This behavior is deprecated.  The package containing that file should instead export the directory containing the file, and you should use rosbuild_include() to include the file explicitly.

  [rosbuild] Including /home/yamokosk/groovy_workspace_old/xacro/cmake/xacro.cmake
  CMake Error at /home/yamokosk/groovy_workspace_old/xacro/cmake/xacro.cmake:12 (message):
    Error occured while trying to execute xacro: Expat parsing error.  Check
    that:

     - Your XML is correctly formed
     - You have the xacro xmlns declaration: xmlns:xacro="http://www.ros.org/wiki/xacro"



    Traceback (most recent call last):

      File "/home/yamokosk/groovy_workspace_old/xacro/xacro.py", line 35, in <module>
        xacro.main()
      File "/home/yamokosk/groovy_workspace_old/xacro/src/xacro.py", line 571, in main
        doc = parse(f)
      File "/usr/lib/python2.7/xml/dom/minidom.py", line 1920, in parse
        return expatbuilder.parse(file)
      File "/usr/lib/python2.7/xml/dom/expatbuilder.py", line 928, in parse
        result = builder.parseFile(file)
      File "/usr/lib/python2.7/xml/dom/expatbuilder.py", line 207, in parseFile
        parser.Parse(buffer, 0)

    xml.parsers.expat.ExpatError: unbound prefix: line 1, column 0

  Call Stack (most recent call first):
    CMakeLists.txt:17 (xacro_add_xacro_file)


  -- Configuring incomplete, errors occurred!
-------------------------------------------------------------------------------}
[ rosmake ] Output from build of package xacro_cmake_bug written to:
[ rosmake ]    /home/yamokosk/.ros/rosmake/rosmake_output-20140207-113232/xacro_cmake_bug/build_output.log                
[rosmake-2] Finished <<< xacro_cmake_bug [FAIL] [ 5.45 seconds ]   
...
```

Correctly get the right error message and see that something was wrong. Then with a correct xacro file:

```
$ rosmake
...
[rosmake-1] Starting >>> xacro [ make ]                                                                                                                                                        
[rosmake-1] Finished <<< xacro [PASS] [ 3.15 seconds ]   
[rosmake-0] Starting >>> xacro_cmake_bug [ make ]                                                                                                                                              
[rosmake-0] Finished <<< xacro_cmake_bug [PASS] [ 5.52 seconds ]
...
$ more sample.urdf 
<?xml version="1.0" ?>
<!-- =================================================================================== -->
<!-- |    This document was autogenerated by xacro from /home/yamokosk/groovy_workspace_old/xacro_cmake_bug/sample.xacro | -->
<!-- |    EDITING THIS FILE BY HAND IS NOT RECOMMENDED                                 | -->
<!-- =================================================================================== -->
<robot name="sim/utility_macros" xmlns:xacro="http://ros.org/wiki/xacro">
  <message> Hello World! </message>
  <message> Hello Universe! </message>
</robot>
```

wjwwood commented:
So I believe you that errors may have not been being raised correctly and that your changes work with the simple example you gave above, but I think that removing the `--deps` argument from the first `execute_process` is not the correct thing to do. If you have a more complicated `xacro` file which has dependencies will likely not work.

I'll try to look into an alternative fix, but right now I don't have time.

yamokosk commented:
I don't see where the `--deps` argument is documented for `execute_process`. Was this an option in CMake 2.4.x? I can only get documentation on CMake's site back to CMake 2.6.x.. there is no mention of this argument in that version.

http://www.cmake.org/cmake/help/cmake2.6docs.html#command:execute_process

yamokosk commented:
oops.. sorry. I see. `--deps` is an argument of xacro.

Commit:
reverted to original xacro.cmake in groovy and added a custom target which seems to correctly hook the building of URDF's from xacros now.

yamokosk commented:
Alright, I think I have fixed it.. again.. ;). For real this time?

I added a custom target `xacro` to the default build target `ALL`. There was a line in the documentation of `add_custom_command`:

http://www.cmake.org/cmake/help/cmake2.6docs.html#command:add_custom_command

> .. The outputs named should be listed as source files in the target for which they are to be generated.

As far as I could tell, the default implementation was not linking the outputs to any existing target. Here is some output from my simple test case. (Note: I did manual builds so that I could more directly see the cmake output instead of having to cut/paste ros build logs).

```
xacro_cmake_bug/build$ cp ../sample.xacro.bad ../sample.xacro
xacro_cmake_bug/build$ make
...
[100%] Generating ../sample.urdf
Expat parsing error.  Check that:
 - Your XML is correctly formed
 - You have the xacro xmlns declaration: xmlns:xacro="http://www.ros.org/wiki/xacro"

Traceback (most recent call last):
  File "/home/yamokosk/groovy_workspace_old/xacro/xacro.py", line 35, in <module>
    xacro.main()
  File "/home/yamokosk/groovy_workspace_old/xacro/src/xacro.py", line 571, in main
    doc = parse(f)
  File "/usr/lib/python2.7/xml/dom/minidom.py", line 1920, in parse
    return expatbuilder.parse(file)
  File "/usr/lib/python2.7/xml/dom/expatbuilder.py", line 928, in parse
    result = builder.parseFile(file)
  File "/usr/lib/python2.7/xml/dom/expatbuilder.py", line 207, in parseFile
    parser.Parse(buffer, 0)
xml.parsers.expat.ExpatError: unbound prefix: line 1, column 0
make[2]: *** [../sample.urdf] Error 1
make[1]: *** [CMakeFiles/xacro.dir/all] Error 2
make: *** [all] Error 2
xacro_cmake_bug/build$ cp ../sample.xacro.good ../sample.xacro
xacro_cmake_bug/build$ make
[100%] Generating ../sample.urdf
[100%] Built target xacro
xacro_cmake_bug/build$ make
[100%] Built target xacro
```

Note that the cmake documentation says that using `add_custom_target` results in the target "ALWAYS CONSIDERED OUT OF DATE". However note in my example output that even though the target is always made, the custom_output is smart enough to know not to regenerate the urdf.

wjwwood commented:
Ok, I'll review these new changes, thanks for updating the pull request. Can I get you to also rebase against master? That will allow the CI to run properly.

rhaschke commented:
@codebot 
I think, this one can be closed. The new conveniency function `xacro_add_files` introduced in #77 will do what was intended by this pull request.

@yamokosk: Using `add_custom_target` in `xacro_add_xacro_file()` will fail on multiple usage of `xacro_add_xacro_file`
